// moons-ta-client/dist/models/events.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MESSAGE_TYPE } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { CoreServer } from './models.js';
import { Tournament } from './models.js';
import { QualifierEvent } from './models.js';
import { Match } from './models.js';
import { User } from './models.js';
// @generated message type with reflection information, may provide speed optimized methods
class Event$Type extends MessageType {
    constructor() {
        super("proto.packets.Event", [
            { no: 1, name: "user_added", kind: "message", oneof: "changedObject", T: () => Event_UserAdded },
            { no: 2, name: "user_updated", kind: "message", oneof: "changedObject", T: () => Event_UserUpdated },
            { no: 3, name: "user_left", kind: "message", oneof: "changedObject", T: () => Event_UserLeft },
            { no: 6, name: "match_created", kind: "message", oneof: "changedObject", T: () => Event_MatchCreated },
            { no: 7, name: "match_updated", kind: "message", oneof: "changedObject", T: () => Event_MatchUpdated },
            { no: 8, name: "match_deleted", kind: "message", oneof: "changedObject", T: () => Event_MatchDeleted },
            { no: 9, name: "qualifier_created", kind: "message", oneof: "changedObject", T: () => Event_QualifierCreated },
            { no: 10, name: "qualifier_updated", kind: "message", oneof: "changedObject", T: () => Event_QualifierUpdated },
            { no: 11, name: "qualifier_deleted", kind: "message", oneof: "changedObject", T: () => Event_QualifierDeleted },
            { no: 12, name: "tournament_created", kind: "message", oneof: "changedObject", T: () => Event_TournamentCreated },
            { no: 13, name: "tournament_updated", kind: "message", oneof: "changedObject", T: () => Event_TournamentUpdated },
            { no: 14, name: "tournament_deleted", kind: "message", oneof: "changedObject", T: () => Event_TournamentDeleted },
            { no: 15, name: "server_added", kind: "message", oneof: "changedObject", T: () => Event_ServerAdded },
            { no: 16, name: "server_deleted", kind: "message", oneof: "changedObject", T: () => Event_ServerDeleted }
        ]);
    }
    create(value) {
        const message = { changedObject: { oneofKind: undefined } };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* proto.packets.Event.UserAdded user_added */ 1:
                    message.changedObject = {
                        oneofKind: "userAdded",
                        userAdded: Event_UserAdded.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.userAdded)
                    };
                    break;
                case /* proto.packets.Event.UserUpdated user_updated */ 2:
                    message.changedObject = {
                        oneofKind: "userUpdated",
                        userUpdated: Event_UserUpdated.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.userUpdated)
                    };
                    break;
                case /* proto.packets.Event.UserLeft user_left */ 3:
                    message.changedObject = {
                        oneofKind: "userLeft",
                        userLeft: Event_UserLeft.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.userLeft)
                    };
                    break;
                case /* proto.packets.Event.MatchCreated match_created */ 6:
                    message.changedObject = {
                        oneofKind: "matchCreated",
                        matchCreated: Event_MatchCreated.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.matchCreated)
                    };
                    break;
                case /* proto.packets.Event.MatchUpdated match_updated */ 7:
                    message.changedObject = {
                        oneofKind: "matchUpdated",
                        matchUpdated: Event_MatchUpdated.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.matchUpdated)
                    };
                    break;
                case /* proto.packets.Event.MatchDeleted match_deleted */ 8:
                    message.changedObject = {
                        oneofKind: "matchDeleted",
                        matchDeleted: Event_MatchDeleted.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.matchDeleted)
                    };
                    break;
                case /* proto.packets.Event.QualifierCreated qualifier_created */ 9:
                    message.changedObject = {
                        oneofKind: "qualifierCreated",
                        qualifierCreated: Event_QualifierCreated.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.qualifierCreated)
                    };
                    break;
                case /* proto.packets.Event.QualifierUpdated qualifier_updated */ 10:
                    message.changedObject = {
                        oneofKind: "qualifierUpdated",
                        qualifierUpdated: Event_QualifierUpdated.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.qualifierUpdated)
                    };
                    break;
                case /* proto.packets.Event.QualifierDeleted qualifier_deleted */ 11:
                    message.changedObject = {
                        oneofKind: "qualifierDeleted",
                        qualifierDeleted: Event_QualifierDeleted.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.qualifierDeleted)
                    };
                    break;
                case /* proto.packets.Event.TournamentCreated tournament_created */ 12:
                    message.changedObject = {
                        oneofKind: "tournamentCreated",
                        tournamentCreated: Event_TournamentCreated.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.tournamentCreated)
                    };
                    break;
                case /* proto.packets.Event.TournamentUpdated tournament_updated */ 13:
                    message.changedObject = {
                        oneofKind: "tournamentUpdated",
                        tournamentUpdated: Event_TournamentUpdated.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.tournamentUpdated)
                    };
                    break;
                case /* proto.packets.Event.TournamentDeleted tournament_deleted */ 14:
                    message.changedObject = {
                        oneofKind: "tournamentDeleted",
                        tournamentDeleted: Event_TournamentDeleted.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.tournamentDeleted)
                    };
                    break;
                case /* proto.packets.Event.ServerAdded server_added */ 15:
                    message.changedObject = {
                        oneofKind: "serverAdded",
                        serverAdded: Event_ServerAdded.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.serverAdded)
                    };
                    break;
                case /* proto.packets.Event.ServerDeleted server_deleted */ 16:
                    message.changedObject = {
                        oneofKind: "serverDeleted",
                        serverDeleted: Event_ServerDeleted.internalBinaryRead(reader, reader.uint32(), options, message.changedObject.serverDeleted)
                    };
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* proto.packets.Event.UserAdded user_added = 1; */
        if (message.changedObject.oneofKind === "userAdded")
            Event_UserAdded.internalBinaryWrite(message.changedObject.userAdded, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.UserUpdated user_updated = 2; */
        if (message.changedObject.oneofKind === "userUpdated")
            Event_UserUpdated.internalBinaryWrite(message.changedObject.userUpdated, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.UserLeft user_left = 3; */
        if (message.changedObject.oneofKind === "userLeft")
            Event_UserLeft.internalBinaryWrite(message.changedObject.userLeft, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.MatchCreated match_created = 6; */
        if (message.changedObject.oneofKind === "matchCreated")
            Event_MatchCreated.internalBinaryWrite(message.changedObject.matchCreated, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.MatchUpdated match_updated = 7; */
        if (message.changedObject.oneofKind === "matchUpdated")
            Event_MatchUpdated.internalBinaryWrite(message.changedObject.matchUpdated, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.MatchDeleted match_deleted = 8; */
        if (message.changedObject.oneofKind === "matchDeleted")
            Event_MatchDeleted.internalBinaryWrite(message.changedObject.matchDeleted, writer.tag(8, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.QualifierCreated qualifier_created = 9; */
        if (message.changedObject.oneofKind === "qualifierCreated")
            Event_QualifierCreated.internalBinaryWrite(message.changedObject.qualifierCreated, writer.tag(9, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.QualifierUpdated qualifier_updated = 10; */
        if (message.changedObject.oneofKind === "qualifierUpdated")
            Event_QualifierUpdated.internalBinaryWrite(message.changedObject.qualifierUpdated, writer.tag(10, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.QualifierDeleted qualifier_deleted = 11; */
        if (message.changedObject.oneofKind === "qualifierDeleted")
            Event_QualifierDeleted.internalBinaryWrite(message.changedObject.qualifierDeleted, writer.tag(11, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.TournamentCreated tournament_created = 12; */
        if (message.changedObject.oneofKind === "tournamentCreated")
            Event_TournamentCreated.internalBinaryWrite(message.changedObject.tournamentCreated, writer.tag(12, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.TournamentUpdated tournament_updated = 13; */
        if (message.changedObject.oneofKind === "tournamentUpdated")
            Event_TournamentUpdated.internalBinaryWrite(message.changedObject.tournamentUpdated, writer.tag(13, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.TournamentDeleted tournament_deleted = 14; */
        if (message.changedObject.oneofKind === "tournamentDeleted")
            Event_TournamentDeleted.internalBinaryWrite(message.changedObject.tournamentDeleted, writer.tag(14, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.ServerAdded server_added = 15; */
        if (message.changedObject.oneofKind === "serverAdded")
            Event_ServerAdded.internalBinaryWrite(message.changedObject.serverAdded, writer.tag(15, WireType.LengthDelimited).fork(), options).join();
        /* proto.packets.Event.ServerDeleted server_deleted = 16; */
        if (message.changedObject.oneofKind === "serverDeleted")
            Event_ServerDeleted.internalBinaryWrite(message.changedObject.serverDeleted, writer.tag(16, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event
 */
export const Event = new Event$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_UserAdded$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.UserAdded", [
            { no: 1, name: "tournament_id", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "user", kind: "message", T: () => User }
        ]);
    }
    create(value) {
        const message = { tournamentId: "" };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string tournament_id */ 1:
                    message.tournamentId = reader.string();
                    break;
                case /* proto.models.User user */ 2:
                    message.user = User.internalBinaryRead(reader, reader.uint32(), options, message.user);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string tournament_id = 1; */
        if (message.tournamentId !== "")
            writer.tag(1, WireType.LengthDelimited).string(message.tournamentId);
        /* proto.models.User user = 2; */
        if (message.user)
            User.internalBinaryWrite(message.user, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.UserAdded
 */
export const Event_UserAdded = new Event_UserAdded$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_UserUpdated$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.UserUpdated", [
            { no: 1, name: "tournament_id", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "user", kind: "message", T: () => User }
        ]);
    }
    create(value) {
        const message = { tournamentId: "" };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string tournament_id */ 1:
                    message.tournamentId = reader.string();
                    break;
                case /* proto.models.User user */ 2:
                    message.user = User.internalBinaryRead(reader, reader.uint32(), options, message.user);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string tournament_id = 1; */
        if (message.tournamentId !== "")
            writer.tag(1, WireType.LengthDelimited).string(message.tournamentId);
        /* proto.models.User user = 2; */
        if (message.user)
            User.internalBinaryWrite(message.user, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.UserUpdated
 */
export const Event_UserUpdated = new Event_UserUpdated$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_UserLeft$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.UserLeft", [
            { no: 1, name: "tournament_id", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "user", kind: "message", T: () => User }
        ]);
    }
    create(value) {
        const message = { tournamentId: "" };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string tournament_id */ 1:
                    message.tournamentId = reader.string();
                    break;
                case /* proto.models.User user */ 2:
                    message.user = User.internalBinaryRead(reader, reader.uint32(), options, message.user);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string tournament_id = 1; */
        if (message.tournamentId !== "")
            writer.tag(1, WireType.LengthDelimited).string(message.tournamentId);
        /* proto.models.User user = 2; */
        if (message.user)
            User.internalBinaryWrite(message.user, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.UserLeft
 */
export const Event_UserLeft = new Event_UserLeft$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_MatchCreated$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.MatchCreated", [
            { no: 1, name: "tournament_id", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "match", kind: "message", T: () => Match }
        ]);
    }
    create(value) {
        const message = { tournamentId: "" };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string tournament_id */ 1:
                    message.tournamentId = reader.string();
                    break;
                case /* proto.models.Match match */ 2:
                    message.match = Match.internalBinaryRead(reader, reader.uint32(), options, message.match);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string tournament_id = 1; */
        if (message.tournamentId !== "")
            writer.tag(1, WireType.LengthDelimited).string(message.tournamentId);
        /* proto.models.Match match = 2; */
        if (message.match)
            Match.internalBinaryWrite(message.match, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.MatchCreated
 */
export const Event_MatchCreated = new Event_MatchCreated$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_MatchUpdated$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.MatchUpdated", [
            { no: 1, name: "tournament_id", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "match", kind: "message", T: () => Match }
        ]);
    }
    create(value) {
        const message = { tournamentId: "" };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string tournament_id */ 1:
                    message.tournamentId = reader.string();
                    break;
                case /* proto.models.Match match */ 2:
                    message.match = Match.internalBinaryRead(reader, reader.uint32(), options, message.match);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string tournament_id = 1; */
        if (message.tournamentId !== "")
            writer.tag(1, WireType.LengthDelimited).string(message.tournamentId);
        /* proto.models.Match match = 2; */
        if (message.match)
            Match.internalBinaryWrite(message.match, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.MatchUpdated
 */
export const Event_MatchUpdated = new Event_MatchUpdated$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_MatchDeleted$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.MatchDeleted", [
            { no: 1, name: "tournament_id", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "match", kind: "message", T: () => Match }
        ]);
    }
    create(value) {
        const message = { tournamentId: "" };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string tournament_id */ 1:
                    message.tournamentId = reader.string();
                    break;
                case /* proto.models.Match match */ 2:
                    message.match = Match.internalBinaryRead(reader, reader.uint32(), options, message.match);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string tournament_id = 1; */
        if (message.tournamentId !== "")
            writer.tag(1, WireType.LengthDelimited).string(message.tournamentId);
        /* proto.models.Match match = 2; */
        if (message.match)
            Match.internalBinaryWrite(message.match, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.MatchDeleted
 */
export const Event_MatchDeleted = new Event_MatchDeleted$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_QualifierCreated$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.QualifierCreated", [
            { no: 1, name: "tournament_id", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "event", kind: "message", T: () => QualifierEvent }
        ]);
    }
    create(value) {
        const message = { tournamentId: "" };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string tournament_id */ 1:
                    message.tournamentId = reader.string();
                    break;
                case /* proto.models.QualifierEvent event */ 2:
                    message.event = QualifierEvent.internalBinaryRead(reader, reader.uint32(), options, message.event);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string tournament_id = 1; */
        if (message.tournamentId !== "")
            writer.tag(1, WireType.LengthDelimited).string(message.tournamentId);
        /* proto.models.QualifierEvent event = 2; */
        if (message.event)
            QualifierEvent.internalBinaryWrite(message.event, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.QualifierCreated
 */
export const Event_QualifierCreated = new Event_QualifierCreated$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_QualifierUpdated$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.QualifierUpdated", [
            { no: 1, name: "tournament_id", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "event", kind: "message", T: () => QualifierEvent }
        ]);
    }
    create(value) {
        const message = { tournamentId: "" };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string tournament_id */ 1:
                    message.tournamentId = reader.string();
                    break;
                case /* proto.models.QualifierEvent event */ 2:
                    message.event = QualifierEvent.internalBinaryRead(reader, reader.uint32(), options, message.event);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string tournament_id = 1; */
        if (message.tournamentId !== "")
            writer.tag(1, WireType.LengthDelimited).string(message.tournamentId);
        /* proto.models.QualifierEvent event = 2; */
        if (message.event)
            QualifierEvent.internalBinaryWrite(message.event, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.QualifierUpdated
 */
export const Event_QualifierUpdated = new Event_QualifierUpdated$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_QualifierDeleted$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.QualifierDeleted", [
            { no: 1, name: "tournament_id", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "event", kind: "message", T: () => QualifierEvent }
        ]);
    }
    create(value) {
        const message = { tournamentId: "" };
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string tournament_id */ 1:
                    message.tournamentId = reader.string();
                    break;
                case /* proto.models.QualifierEvent event */ 2:
                    message.event = QualifierEvent.internalBinaryRead(reader, reader.uint32(), options, message.event);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string tournament_id = 1; */
        if (message.tournamentId !== "")
            writer.tag(1, WireType.LengthDelimited).string(message.tournamentId);
        /* proto.models.QualifierEvent event = 2; */
        if (message.event)
            QualifierEvent.internalBinaryWrite(message.event, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.QualifierDeleted
 */
export const Event_QualifierDeleted = new Event_QualifierDeleted$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_TournamentCreated$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.TournamentCreated", [
            { no: 1, name: "tournament", kind: "message", T: () => Tournament }
        ]);
    }
    create(value) {
        const message = {};
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* proto.models.Tournament tournament */ 1:
                    message.tournament = Tournament.internalBinaryRead(reader, reader.uint32(), options, message.tournament);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* proto.models.Tournament tournament = 1; */
        if (message.tournament)
            Tournament.internalBinaryWrite(message.tournament, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.TournamentCreated
 */
export const Event_TournamentCreated = new Event_TournamentCreated$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_TournamentUpdated$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.TournamentUpdated", [
            { no: 1, name: "tournament", kind: "message", T: () => Tournament }
        ]);
    }
    create(value) {
        const message = {};
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* proto.models.Tournament tournament */ 1:
                    message.tournament = Tournament.internalBinaryRead(reader, reader.uint32(), options, message.tournament);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* proto.models.Tournament tournament = 1; */
        if (message.tournament)
            Tournament.internalBinaryWrite(message.tournament, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.TournamentUpdated
 */
export const Event_TournamentUpdated = new Event_TournamentUpdated$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_TournamentDeleted$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.TournamentDeleted", [
            { no: 1, name: "tournament", kind: "message", T: () => Tournament }
        ]);
    }
    create(value) {
        const message = {};
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* proto.models.Tournament tournament */ 1:
                    message.tournament = Tournament.internalBinaryRead(reader, reader.uint32(), options, message.tournament);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* proto.models.Tournament tournament = 1; */
        if (message.tournament)
            Tournament.internalBinaryWrite(message.tournament, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.TournamentDeleted
 */
export const Event_TournamentDeleted = new Event_TournamentDeleted$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_ServerAdded$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.ServerAdded", [
            { no: 1, name: "server", kind: "message", T: () => CoreServer }
        ]);
    }
    create(value) {
        const message = {};
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* proto.models.CoreServer server */ 1:
                    message.server = CoreServer.internalBinaryRead(reader, reader.uint32(), options, message.server);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* proto.models.CoreServer server = 1; */
        if (message.server)
            CoreServer.internalBinaryWrite(message.server, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.ServerAdded
 */
export const Event_ServerAdded = new Event_ServerAdded$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Event_ServerDeleted$Type extends MessageType {
    constructor() {
        super("proto.packets.Event.ServerDeleted", [
            { no: 1, name: "server", kind: "message", T: () => CoreServer }
        ]);
    }
    create(value) {
        const message = {};
        globalThis.Object.defineProperty(message, MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* proto.models.CoreServer server */ 1:
                    message.server = CoreServer.internalBinaryRead(reader, reader.uint32(), options, message.server);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* proto.models.CoreServer server = 1; */
        if (message.server)
            CoreServer.internalBinaryWrite(message.server, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message proto.packets.Event.ServerDeleted
 */
export const Event_ServerDeleted = new Event_ServerDeleted$Type();
